### PR TITLE
User Management support

### PR DESCRIPTION
**Pull request by amadeo-workos**
## Description
Support for all User Management APIs.

### CODE REVIEW

File: src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithCodeOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithCodeOptionsBuilder.kt:7]: AFAICT, this constructs the options to be sent in the body of the request rather than a URL. Maybe this?

```suggestion
 * Builder for options when authentication with an authorization code.
```

File: src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithMagicAuthOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithMagicAuthOptionsBuilder.kt]: Similar comment:

```suggestion
 * Builder for options when authenticating with Magic Auth.
```

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt:7]:
```suggestion
 * Builder for options when authenticating with MFA.
 ```

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt:11]:
```suggestion
 * @param code The one-time code generated by the user's second-factor device.
 ```

File: src/main/kotlin/com/workos/user_management/builders/CreateOrUpdateUserOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/CreateOrUpdateUserOptionsBuilder.kt:9]: `email` is only a possible option when creating a user. Should we have separate objects for create and update to make sure people don't attempt to pass `email` when updating a user?

Comment (mthadley) [reply src/main/kotlin/com/workos/user_management/builders/CreateOrUpdateUserOptionsBuilder.kt:9]: Might make sense to have some kind of `BaseCreateOrUPdateUserOptions` class that both can inherit from with only the common options, but I'll defer to you as the more experienced Kotlin developer.

File: src/main/kotlin/com/workos/user_management/builders/EnrolledAuthenticationFactorOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/EnrolledAuthenticationFactorOptionsBuilder.kt:6]:
```suggestion
 * Builder for options when enrolling an MFA factor.
```

File: src/main/kotlin/com/workos/user_management/builders/ListOrganizationMembershipsOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/ListOrganizationMembershipsOptionsBuilder.kt:7]:
```suggestion
 * Builder for options when listing OrganizationMemberships.
```

File: src/main/kotlin/com/workos/user_management/models/Authentication.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/models/Authentication.kt:28]:
```suggestion
  @JsonProperty("impersonator")
```

File: src/main/kotlin/com/workos/user_management/models/AuthenticationChallenge.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/models/AuthenticationChallenge.kt:9]: We document the `object` param here, but I don't see it in the declaration below. Same for several other objects.

Is there but derived somehow elsewhere, or are we missing some code to deserialize it?

File: src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt:8]:
```suggestion
 * A list of authentication Factors.
```

File: src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithCodeOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithCodeOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithMagicAuthOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/usermanagement/builders/AuthenticationWithMagicAuthOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/AuthenticationWithTotpOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/builders/CreateOrUpdateUserOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/CreateOrUpdateUserOptionsBuilder.kt]: OMG good call... for some reason I read from the docs that both create and update were operating off of the email. I just made the change, and also pushed an `AbstractUserOptionsBuilder<OptionsObject>` to simplify the code of both create/update builders.

File: src/main/kotlin/com/workos/user_management/builders/EnrolledAuthenticationFactorOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/EnrolledAuthenticationFactorOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/builders/ListOrganizationMembershipsOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/ListOrganizationMembershipsOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/models/Authentication.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/models/Authentication.kt]: Great catch!! 🔥

File: src/main/kotlin/com/workos/user_management/builders/UpdateOrganizationMembershipOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/UpdateOrganizationMembershipOptionsBuilder.kt]: 👍

File: src/main/kotlin/com/workos/user_management/models/AuthenticationChallenge.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/models/AuthenticationChallenge.kt]: Removed those comments, not needed anymore.

File: src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt]: 👍

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithMagicAuthOptionsBuilder.kt
Comment (mthadley) [inline src/main/kotlin/com/workos/user_management/builders/AuthenticationWithMagicAuthOptionsBuilder.kt:7]: This should be "Magic Auth", since it is in `AuthenticationWithMagicAuthOptionsBuilder`, right?

```suggestion
 * Builder for options when authenticating with Magic Auth.
```

File: src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt
Comment (mthadley) [reply src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt]: Oh, wait. Do these brackets actually mean something as far as these doc comments go? As in, would they link to the type-definitions of the relevantly named object if say a doc generator to HTML was run?

I had assumed these were copy-paste mistakes, but now realizing that may have been wrong.

Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/models/AuthenticationFactors.kt]: Yes. They also provide the full documentation of the type when hovering over in IntelliJ, which can be useful in those cases were the structure of the class is not trivial:

image

While it is not a big deal at the moment, as we don't rely on any autogened documentation for the SDKs yet, I went ahead and edited the documentation of the non-literal types found in the model classes.

File: src/main/kotlin/com/workos/user_management/builders/AuthenticationWithMagicAuthOptionsBuilder.kt
Comment (amadeo-workos) [reply src/main/kotlin/com/workos/user_management/builders/AuthenticationWithMagicAuthOptionsBuilder.kt]: 👍

**Review by mthadley — DISMISSED** [review]

**Comment by amadeo-workos** [comment]
@mthadley Can I have another :stamp:? I pushed an extra commit to just rename the name of the folder/package? Thanks!

**Review by mthadley — APPROVED** [review]

**Comment by amadeo-workos** [comment]
@PaulAsjes are you around to approve this? It was originally approved by @mthadley but I had to rebase my branch and dismissed their approval. Thanks!

**Comment by amadeo-workos** [comment]
Sigh... no idea why it is not letting me merge... why 2 commits are verified but 1 is not 👎

**Comment by amadeo-workos** [comment]
@ibrahimbutt

The UM support in the Java SDK has shipped! :boat:  it is under the `3.1.0-beta.user-management1` version. Please check it out and let us know how it goes!

Thanks!

**Comment by ibrahimbutt** [comment]
@amadeo-workos trying something new here. Are you hiring? I have 5-6 years of experience in FinTech using Java and Spring Boot (API Design and Development).

**Pull request merged**
